src/crimson/osd/ops_executer.cc

```diff
@@ -100,51 +100,219 @@ seastar::future<> OpsExecuter::do_op_call(OSDOp& osd_op)
-                      const std::string& nspace,
-                      uint64_t limit)
+
+seastar::future<bool, hobject_t> OpsExecuter::pgls_filter(
+  /*const TODO: fix get_xattr?*/ PGLSFilter& filter,
+  const hobject_t& sobj)
 {
-  hobject_t lower_bound;
-  try {
-    ceph::decode(lower_bound, indata);
-  } catch (const buffer::error& e) {
-    throw std::invalid_argument("unable to decode PGNLS handle");
+  if (const auto xattr = filter.get_xattr(); !xattr.empty()) {
+    logger().debug("pgls_filter: filter is interested in xattr={} for obj={}",
+                   xattr, sobj);
+    return backend.getxattr(sobj, xattr).then_wrapped(
+      [&filter, sobj] (auto futval) {
+        logger().debug("pgls_filter: got xvalue for obj={}", sobj);
+
+        ceph::bufferlist val;
+        if (!futval.failed()) {
+          val.push_back(std::move(futval).get0());
+        } else if (filter.reject_empty_xattr()) {
+          return seastar::make_ready_future<bool, hobject_t>(false, sobj);
+        }
+        const bool filtered = filter.filter(sobj, val);
+        return seastar::make_ready_future<bool, hobject_t>(filtered, sobj);
+    });
+  } else {
+    ceph::bufferlist empty_lvalue_bl;
+    const bool filtered = filter.filter(sobj, empty_lvalue_bl);
+    return seastar::make_ready_future<bool, hobject_t>(filtered, sobj);
   }
+}
+
+seastar::future<ceph::bufferlist>
+OpsExecuter::do_pgnls_common(
+  const hobject_t& lower_bound,
+  const std::string& nspace,
+  uint64_t limit,
+  PGLSFilter* filter)
+{
   const auto pg_start = pg.get_pgid().pgid.get_hobj_start();
-  const auto pg_end = pg.get_pgid().pgid.get_hobj_end(pg.get_pool().info.get_pg_num());
+  auto pg_end = pg.get_pgid().pgid.get_hobj_end(pg.get_pool().info.get_pg_num());
   if (!(lower_bound.is_min() ||
         lower_bound.is_max() ||
         (lower_bound >= pg_start && lower_bound < pg_end))) {
     // this should only happen with a buggy client.
     throw std::invalid_argument("outside of PG bounds");
   }
+
   return backend.list_objects(lower_bound, limit).then(
-    [lower_bound, pg_end, nspace](auto objects, auto next) {
-      auto in_my_namespace = [&nspace](const hobject_t& o) {
+    [this, filter, nspace](auto objects, auto next) {
+      auto in_my_namespace = [&nspace](const hobject_t& obj) {
         using ceph::common::local_conf;
-        if (o.get_namespace() == local_conf()->osd_hit_set_namespace) {
+        if (obj.get_namespace() == local_conf()->osd_hit_set_namespace) {
           return false;
         } else if (nspace == librados::all_nspaces) {
           return true;
         } else {
-          return o.get_namespace() == nspace;
+          return obj.get_namespace() == nspace;
+        }
+      };
+      auto to_pglsed = [this, filter] (const hobject_t& obj) {
+        // this transformation looks costly. However, I don't have any
+        // reason to think PGLS* operations are critical for, let's say,
+        // general performance.
```

Comment:
yeah, another way is to use `seastar::map_reduce()`, to 1) save the effort to filter the already filtered objects 2) avoid the space to keep the `tuple<bool, object>` even if the object is filtered out.

Reply:
Yup, I'm going to extend the comment with the idea.

Reply:
Added the comment.